### ISSUE TITLE
Does not work with NPM workspaces

### ISSUE THREAD

**Issue by ghost** [electron-userland/electron-builder#7103]
* **Electron-Builder Version**: 23.3.3

* **Node Version**: v18.3.0

* **Electron Version**: 20.1.0

* **Electron Type (current, beta, nightly)**: current

* **Target**:  win32 x64

electron builder does not work in npm workspace
it deletes app-builder-bin during "installing production dependencies" and then throws error:
```
• packaging       platform=win32 arch=x64 electron=20.1.0 appOutDir=dist\win-unpacked
The system cannot find the path specified.
  ⨯ spawn C:\Users\user\dev\test-workspace\node_modules\app-builder-bin\win\x64\app-builder.exe ENOENT
```
 also it deletes more packages from node_modules, so next run throws
 ```
'electron-builder' is not recognized as an internal or external command
```
Reproduction:
` npm run build -w a` in [test-workspace.zip](https://github.com/electron-userland/electron-builder/files/9458943/test-workspace.zip)

**Comment by paulantoine2** [comment]
For me, i just get stuck at `installing production dependencies`but the behavior stay the same.

However, it's working with electron-builder version 22 !

**Comment by RadoslavK** [comment]
I am having the same or similar problem and found it started happening with npm 8.5.0 (node 16.14.1) because when I downgrade to 8.3.1 (that came with node 16.14.1) while keeping a higher node version, it works. Npm changelog says something about workspaces but I am not sure if its related.

**Comment by Hinaser** [comment]
Try to add build option below in your `package.json` which has `main` entry for electron app:
```javascript
{
  "main": "build/electron/main.js",
  ...,
  "build": {
    ...,
    "npmRebuild": false
  }
}
```

**Comment by joshtynjala** [comment]
Confirmed that setting `npmRebuild` to `false` works for my project, but I suspect it is not a universal workaround.

It looks like this code is where the _installing production dependencies_ happens:

https://github.com/electron-userland/electron-builder/blob/01c679107435c6afd0b2de8c44d3f140d20c5577/packages/app-builder-lib/src/util/yarn.ts#L86-L120

It seems to be running `node install --production`. I'm guessing that this command is deleting dependencies needed by electron-builder, including the mentioned **app-builder-bin**.

**Comment by nanamicat** [comment]
setting npmRebuild to false works for me, but I don't think it's a suitable solution.
another problem is when depend on a non-exact electron version like `^23.1.0`, it can't find installed electron package and will throw.

**Comment by jkroepke** [comment]
I look into the code and figure out that create an empty node_module directory on the specific project where electron-builds solve my issue here. npm rebuild works fine in my case.

See https://github.com/MuhammedKalkan/OpenLens/blob/c4b099462f9e4cdf405d02993a40e5563278bfd1/.github/workflows/main.yml#L47-L49 for a living example.

**Comment by rotu** [comment]
.

**Comment by jkroepke** [comment]
still an issue

**Comment by st0rm348** [comment]
Yes, still an issue: having to create an empty 'node_modules' folder on every workspace and having to fix the version of Electron on `package.json` is very annoying...

**Comment by mmaietta** [comment]
Very intereesting.

Is there a proper way to detect if we're in a npm workspace and can discern that dependencies are already installed?

**Comment by vincesp** [comment]
> Very intereesting.
> 
> Is there a proper way to detect if we're in a npm workspace and can discern that dependencies are already installed?

[`getElectronVersionFromInstalled(projectDir)`](https://github.com/electron-userland/electron-builder/blob/2ca621289e025a6d0856c7c588aaac40d267cb5f/packages/app-builder-lib/src/electron/electronVersion.ts#L26) is trying to implement its own package resolution algorithm which probably is not a good idea. A better solution could be:

```
cd ${projectDir}
node -p "require.resolve('electron')"
```

(…or `node -p "import.meta.resolve('electron')"` if project is a module.)

**Comment by vincesp** [comment]
It is important that the `resolve()` takes place in the context of the electron project, not in the context of the electron-builder itself.

**Comment by rotu** [comment]
not stale

**Comment by rotu** [comment]
still not stale

**Comment by scottcorgan** [comment]
Yes, still very much a difficult issue

**Comment by vincesp** [comment]
Implementation of `getElectronVersionFromInstalled()` still needs to be fixed.

**Comment by scriptcoded** [comment]
For anyone looking for a workaround, here's the summary.

1. You need to pin your version of `electron`. Only this package has to be pinned. Example:
    ```json
    "dependencies": {
      "electron": "31.4.0",
      "electron-builder": "^24.13.3"
    }
    ```
2. You need to have an empty `node_modules` folder in your package directory, otherwise NPM hangs during install (I have no clue why). An easy way of making sure this is always the case is to use a preinstall script in your package.json for the package which uses electron:
    ```json
    {
      "scripts": {
        "preinstall": "mkdir node_modules"
      }
    }
    ```
    And if you need this to work cross-platform (looking at you Windows) you could use something like [shx](https://www.npmjs.com/package/shx):
    ```json
    {
      "scripts": {
        "preinstall": "shx mkdir node_modules"
      }
    }
    ```

**Comment by rotu** [comment]
How is this fixed by #8715?

Afaict, workspace packages still seem to get symlinked from the `node_modules` directory and so seem to be broken in the resulting asar.
